Assets/Scripts/BoardManager.cs

```diff
@@ -75,14 +218,129 @@ Vector3 RandomPosition()
         return RandomPosition;
     }
 
-    void LayoutObjectAtRandom(GameObject[] tileArray)
+    Vector3 RandomPositionNotCloseToInnerWall(List<Vector3> innerWallGridPositions)
     {
-        for (int i = 0; i < wallCount; i++)
+        int randomIndex = Random.Range(0, gridPositions.Count);
+        Vector3 RandomPosition = gridPositions[randomIndex];
+        if (!CheckIfPositionIsNextToPositions(RandomPosition, innerWallGridPositions))
         {
-            Vector3 randomPosition = RandomPosition();
-            GameObject tileChoice = tileArray[Random.Range(0, tileArray.Length)];
-            Instantiate(tileChoice, randomPosition, Quaternion.identity);
+            gridPositions.RemoveAt(randomIndex);
+            return RandomPosition;
+        }
+        return RandomPositionNotCloseToInnerWall(innerWallGridPositions);
+    }
+
+    bool CheckIfPositionIsNextToPositions(Vector3 position, List<Vector3> positions)
+    {
+        foreach (var pos in positions)
+        {
+            int diffX = (int)Math.Abs((float)(pos.x - position.x));
+            int diffY = (int)Math.Abs((float)(pos.y - position.y));
+
+            if (diffX + diffY <= 2 && diffX != 2 && diffY != 2)
+                if (diffX + diffY > 0)
+                    return true;
         }
+        return false;
+    }
+
+    bool CheckIfPositionIsNextToWall(Vector3 position, List<Vector3> innerWalls)
+    {
+        return (CheckIfPositionIsNextToPositions(position, outerWallGridPositions) || CheckIfPositionIsNextToPositions(position, innerWallGridPositions)) && !CheckIfPositionIsNextToPositions(position, innerWalls);
+    }
+
+    void LayoutWallAtRandom()
+    {
+        for (int i = 0; i < innerWallCount; i++)
+        {
+            List<Vector3> innerWalls = new List<Vector3>();
+            bool isTouchingWall = false;
+            Vector3 startingPosition = RandomPositionNotCloseToInnerWall(innerWallGridPositions);
+
+            GameObject tileChoice = outerWallTiles[Random.Range(0, outerWallTiles.Length)];
+            Instantiate(tileChoice, startingPosition, Quaternion.identity);
+
+            innerWallGridPositions.Add(startingPosition);
+            innerWalls.Add(startingPosition);
+            if (CheckIfPositionIsNextToWall(startingPosition, innerWallGridPositions))
+            {
+                isTouchingWall = true;
+            }
+
+            Vector3 randomPosition = startingPosition;
+
+            for (int a = 0; a < innerWallLength - 1; a++)
+            {
+                List<Vector3> possiblePositions = new List<Vector3>() {
+                    new Vector3(randomPosition.x + 1, randomPosition.y, 0),
+                    new Vector3(randomPosition.x, randomPosition.y + 1, 0),
+                    new Vector3(randomPosition.x - 1, randomPosition.y, 0),
+                    new Vector3(randomPosition.x, randomPosition.y -1, 0),
+                    };
+                foreach (var item in innerWalls)
```

Comment:
You could gain a lot of time by storing a 2 dimensional array representing your grid. 
In each cell you could just use an enum to tells whether it is a wall, floor, an enemy or nothing (of border)

Then just make a method IsCellEmpty(Vector2 position) returning true when the corresponding cell has the floor enum in it.